lib/Sema/TypeCheckAttr.cpp

```diff
@@ -4749,11 +4749,24 @@ static bool typeCheckDerivativeAttr(ASTContext &Ctx, Decl *D,
   // to be enabled.
   if (checkIfDifferentiableProgrammingEnabled(Ctx, attr, D->getDeclContext()))
     return true;
+
   auto *derivative = cast<FuncDecl>(D);
   auto originalName = attr->getOriginalFunctionName();
 
   auto *derivativeInterfaceType =
-      derivative->getInterfaceType()->castTo<AnyFunctionType>();
+      derivative->getInterfaceType()->getAs<AnyFunctionType>();
+  if (!derivativeInterfaceType) {
+    // Crashes if the instance is not an ErrorType.
+    if (derivative->getInterfaceType()->is<ErrorType>()) {
+      diags.diagnose(attr->getLocation(),
+                     diag::derivative_attr_unknown_error);
+      return true;
+    } else {
+      // This should never happen, but leaving a runtime diagnostic incase it
+      // does because little is known about this failure.
+      llvm::report_fatal_error("Unknown failure in typeCheckDerivativeAttr.");
+    }
+  }
```

Comment:
If we encounter an `ErrorType` here, it means we already produced an appropriate diagnostic elsewhere. I think we should simply bail out if there's error in the interface type. Also, we should not be expecting `getInterfaceType()` to return a null type.

Reply:
Calling `Decl->isInvalid()` returns false, so guarding against that does nothing for bailing out early. The only way to do so is to check that the pointer is null, then exit early. Since this behavior deserves to be caught, I'm issuing a second diagnostic.

Reply:
`getInterfaceType()` is not what's returning a null. It's `getAs<...>` that's returning the null. In fact, I call a function on the result of `getInterfaceType()` in that conditional block.

Reply:
Are you sure? `Decl->isInvalid()` on a `FuncDecl` is equivalent to `getInterfaceType()->hasError()`. It *should* have an error with the code in your example.

Reply:
So just check `hasError()`, silently bail if true, then cast to `AnyFunctionType` as was previously done?

Reply:
Sounds good. You can avoid defining an additional variable by checking the declaration instead (using `Decl->isInvalid()`).

Reply:
I just said that Decl->isInvalid doesn't do anything useful. It always returns true here.

Reply:
> It always returns true here

That is exactly what you need. The declaration is invalid when its interface type contains an `ErrorType`.

Reply:
I edited my response.

Reply:
Can you show me the exact code you're running?

Reply:
The behavior just changed. It now does return true. Maybe I misremembered my experience debugging it.